### ISSUE TITLE
Android crashes intermittently when adding Firebase document to collection

### ISSUE THREAD

**Issue by garymich** [myflashlab/Firebase-ANE#246]
## AnnimateCC##
* Android API Version: AIR 30..0.0.107
* iOS Version: N/A
* AIR SDK Version: Adobe AIR 30.0 SDK

## Problem Description ##
Odd that this is happening on only one collection. The other collections we use we have no issues except this one. In debug when I click the collection add method the app freezes, the listener for the ```FirestoreEvents.COLLECTION_ADD_SUCCESS``` and ```FirestoreEvents.COLLECTION_ADD_FAILURE``` never execute. After a short presiod of time the app just ends. Oddly enough when I check the database the new document did in fact get created. I am attaching the Main.as file and the manifest, the problem occurs in the function: ```createEventPrizeTicket``` which starts at line 353. Any help you could provide would be greatly appreciated as we are under a tight deadline to get this product delivered.

[ZillyWin-app.zip](https://github.com/myflashlab/Firebase-ANE/files/2408282/ZillyWin-app.zip)

**Comment by myflashlab** [comment]
Please post the crash logs for us to be able to study the problem in more details. Thanks.

**Comment by garymich** [comment]
Despite searching on Google I was not able to find any information on where to find these files. I am using Windows 10 - where should I look?

**Comment by garymich** [comment]
I tried to update to V7.0.0 but encountered too many issues with the changes to Firebase Authentication and therefore reverted back to V6.5.0. Is there a way I could only update the Firestore ANE to V7.0.0?

I am attaching the logcat files as requested and hope you will be able to help as we have to demo the product for the client on Tuesday morning. Thank you for all your help.
[AddLogcatOuttput.txt](https://github.com/myflashlab/Firebase-ANE/files/2453624/AddLogcatOuttput.txt)
[UpdateLogcatOutput.txt](https://github.com/myflashlab/Firebase-ANE/files/2453625/UpdateLogcatOutput.txt)

**Comment by myflashlab** [comment]
Thanks for the error logs. What I noticed there is that the app closes due to being frozen for a few seconds and there's no native crash logs. This means the AIR side is freezing the app and is unresponsive for a few seconds and eventually force closes the app.

Considering that you said this happens only on this specific collection, maybe there's something special about this branch? right? can you duplicate the collection and test if it would work with less data in the branch? maybe when you are receiving the data from database, you are doing too much work with it that causes the app to freeze?

> Is there a way I could only update the Firestore ANE to V7.0.0?

No, it's either all or none! Anyway, upgrading to V7.0.0 is super easy when you are using ANELAB. Yet, you should spend a few minutes and get ready your FirebaseConsole for [DynamicLinks](https://github.com/myflashlab/Firebase-ANE/wiki/I.1-Add-Dynamic-links) and [Invites](https://github.com/myflashlab/Firebase-ANE/wiki/J.1-Add-Invites).

It is possible that the problem would be fixed automatically because Firestore on Google Firebase is still in beta and they change a LOT of things with each release.

Yet, if you don't want to upgrade just yet (which I suggest you should) try checking your collection more and see what is saved in it differently? because when I looked at the logs, this does not seem like a native side problem. It feels like the problem is either on the AIR and ActionScript side or even maybe on the data being saved on the server side.

**Comment by garymich** [comment]
Hello, thank you for your help! I thought I would try once more to get V7 working and found I had missed something in the dynamic link I created in the Firebase console which I ave fixed. When I try to debug the application I now get the following error: Attempting to launch and connect to Player using URL C:/ZillyWin/ZillyWin-app.xml
Debug session terminated.
The content cannot be loaded because there was a problem loading an extension: Error: Duplicate platform Android-ARM found for extension com.myflashlab.air.extensions.dependency.firebase.common

Any idea what I have done wrong? The xml file was generated by the ANELAB application and I have not changed it. Thanks for you help :)

**Comment by garymich** [comment]
Please disregard the above as when I tried to debug to my android I did not get the error. Unfortunately, I am still not able to get the authentication to work. I am rechecking your sample code to see if I ave missed something but do not believe I have.

**Comment by myflashlab** [comment]
Make sure you’re reading the wiki to know how dynamicLinks and invites are setup on Firebase Console also.

In case you need live help via Skype/teamviewer, feel free to contact us through https://www.myflashlabs.com/contact/

**Comment by garymich** [comment]
Hi, I did read the wiki and believe I have the dynamic link setup correctly. Do I need to setup invites?

**Comment by garymich** [comment]
I have checked my code and believe I have made the required changes in order to be able to sign in using email address and password but when I try it on my phone nothing happens after the line: Auth.signIn(authProvider.getCredential());
I am attaching the as3 file as perhaps I am still missing something in that regard.
[Main.zip](https://github.com/myflashlab/Firebase-ANE/files/2454129/Main.zip)

**Comment by garymich** [comment]
I appreciate that I can contact you with Skype. How would we arrange this? I suppose I would like to confirm that I have my dynamic link properly configured in the Firebase console. I am attaching 2 screenshots of the dynamic link configuration. The behavior for both Android & IOS is to open the hoe page of the Zilly.win website. I have tested it and t does workfrom the console. In all honesty, I have absolutely no idea why we even need this for authentication to work. Please let me know if I can provide you with any further information and once again thank you for all your help. :)
![dl1](https://user-images.githubusercontent.com/41267475/46584744-f7d76a00-ca34-11e8-8210-a8231671eaed.png)
![dl2](https://user-images.githubusercontent.com/41267475/46584745-fad25a80-ca34-11e8-8cbf-b1bacd1b6a09.png)

**Comment by myflashlab** [comment]
I checked your .as file, and noticed that you have ```Firebase.init();``` but in [wiki docs](https://github.com/myflashlab/Firebase-ANE/wiki/I.2-Init-Dynamic-links), we have clearly mentioned that you have to pass true when initializing the FirebaseCore

```actionscript
Firebase.init(true);
```

> In all honesty, I have absolutely no idea why we even need this for authentication to work

You should ask this from Firebase team for more details but if you ask us, I must say that Firebase Authentication has many different features which depends on dynamiclinks and invitations.

Are you sure you have set the manifest correctly? Are you using ANELAB, right?

> I appreciate that I can contact you with Skype. How would we arrange this?

As I said, you need to arrange this with my colleagues through our contact form. https://www.myflashlabs.com/contact/

**Comment by garymich** [comment]
Hi, sorry I messed that one. I made that change but the init method now fails saying the config file cannot be found. This never happened before and whatever file it is looking for is in the same place as it always has been. What file is it looking for and could it possibly be looking for a different file name after having implemented the dynamic link? Thanks again for all your help!

**Comment by garymich** [comment]
Hi, sorry to be a bother. I am looking at your API documentation and from reading it I would presume that using DynamicLinks in my project is optional. The above error I reported is likely because I do not have: <key>com.myflashlabs.firebase.deepLinkURLScheme</key>
         <string>YourCustomSchemeString</string>
in my manifest file. Should I be forced to use DynamicLinks juest in order to get authentication to work then what value would I use for YourCustomSchemeString in the above tag and where would I put that tag in my manifest file? More importantly, could I not include the DynamicLinks ANE in my project change the code back to what it was before and authenticate as I did before?

**Comment by garymich** [comment]
Hi can I get your help on how to configure deepLinkURLScheme key in my app.xml file? I am surprised that this was not done y the ANELAB program as I used it to import the ANE into my project. Please help at your earliest convenience. Thank you for all your help :)

**Comment by myflashlab** [comment]
> init method now fails saying the config file cannot be found

This happens when ```GoogleService-Info.plist``` is not found. make sure you have that added in your bundle.

> could it possibly be looking for a different file name after having implemented the dynamic link?

Nope, it has nothing to do with that. you just need to make sure the file is packaged in your app.

> The above error I reported is likely because I do not have: com.myflashlabs.firebase.deepLinkURLScheme
YourCustomSchemeString
in my manifest file

Nope, not related as explained above. However, about ```com.myflashlabs.firebase.deepLinkURLScheme```, well, the Firebase docs is not very clear about it. here's a similar question that a native developer asked: https://stackoverflow.com/questions/43749819/what-is-customurlscheme-in-firebase-dynamic-links

leave it as is for now:
```xml
<key>com.myflashlabs.firebase.deepLinkURLScheme</key>
<string>dlscheme</string>
```

> and where would I put that tag in my manifest file?

Always use the sample project to know where things are placed: https://github.com/myflashlab/Firebase-ANE/blob/master/AIR/Main-app.xml#L258

If you're using ANELAB, it is putting everything in the right place automatically. The other important settings about DynamicLinks are:

1. host scheme url which is usually your website url: https://github.com/myflashlab/Firebase-ANE/blob/master/AIR/Main-app.xml#L49-L52

```xml
<!-- put your own website name instead of myflashlabs.com -->
<!--<data android:scheme="http" android:host="www.myflashlabs.com" />
<data android:scheme="https" android:host="www.myflashlabs.com" />
```

2. For Firebase Invites, you will need to set the scheme for the iOS again like this: https://github.com/myflashlab/Firebase-ANE/blob/master/AIR/Main-app.xml#L234-L250

```xml
		<!-- 
			Required by firebaseInvites.ane
			Change google client id to your own. Find it in GoogleService-Info.plist
		-->
		<key>CFBundleURLTypes</key>
		<array>
			<dict>
				<key>CFBundleTypeRole</key>
				<string>Editor</string>
				<key>CFBundleURLName</key>
				<string>google</string>
				<key>CFBundleURLSchemes</key>
				<array>
					<string>com.googleusercontent.apps.163763113291-5j7w6ara6rgafjub3636in7h8q2lgbeb</string>
				</array>
			</dict>
		</array>
```

```xml
<!-- Required by firebaseDynamicLinks.ane -->
<key>com.myflashlabs.firebase.deepLinkURLScheme</key>
<string>dlscheme</string>
```

And finally, under the ```Entitlements``` tag, you should add your **app_code** which is found in your Firebase Console: https://github.com/myflashlab/Firebase-ANE/blob/master/AIR/Main-app.xml#L294-L304

```xml
<key>com.apple.developer.associated-domains</key>
<array>
    <string>applinks:e73ko.app.goo.gl</string>
</array>
```

**NOTICE:** make sure you are regenerating your iOS provision file. explained in more details [here](https://github.com/myflashlab/Firebase-ANE/wiki/I.2-Init-Dynamic-links#you-need-a-new-ios-provision-file)

**NOTICE:** Also when adding Firebase Invites, make sure you have added your app SHA fingerprints to Firebase console. [fine more details here](https://github.com/myflashlab/Firebase-ANE/wiki/J.1-Add-Invites#connect-your-adobeair-app-to-firebase)

Basically I am repeating all the information found wiki pages and different tickets. though, I hope you will find this information more useful as it's been organized for your specific need. If you have more questions, do ask.

**Comment by garymich** [comment]
Hello, thank you for your help and listing it all in one place which makes it much easier. 

Regarding: This happens when GoogleService-Info.plist is not found. make sure you have that added in your bundle.

How/where do I set this and has this changed because of the version changes? We have never had an issue with it finding this file before. I think if you can help me with this last item I might be able to get it to work.

**Comment by myflashlab** [comment]
No, it has not changed. This is again explained in details at wiki: https://github.com/myflashlab/Firebase-ANE/wiki/A.2.-Add-Firebase-to-your-app

> copy the final config files you downloaded from Firebase console to the assets folder of your AIR project. (Notice: By the assets folder, it's meant the folder in your app where your main app .swf file is. These files must be available in this location: File.applicationDirectory. You need to make sure these config files are available in your final .apk or .ipa builds.)

**Comment by garymich** [comment]
Sorry, I did figure that one out on my own. Oddly enough I/we never had to include this file in the Target/Include files before. I did that and I no longer get that error but it still does not authenticate so I have to think something is still not configured correctly. Unfortunately, I cannot spend anymore time presently to try to make the new version work so will need to revert back to V6.5 and see if I can come up with a workaround to read/write the data for the collection that is failing. If I can I will revisit upgrading to V7.0 if not it won't matter as well will likely lose the contract. Thank you again for all of your help - it is greatly appreciated! :)

**Comment by myflashlab** [comment]
Wish you the best on your contract. yeah, about using the current version 6.5.0, I really think there's something on the server side or your AIR side which makes the app unresponsive for several seconds... it kinda feels like it is unlikely that the source of the problem would be the ANE or the Firestore SDK. because you said it's working on some collections but does not on some others!

**Comment by garymich** [comment]
Thanks for your best wishes and all your help. After a lot of different attempts I create a new collection defined all the fields to be strings and it actually works. Why I don't know but at least we ca  move forward for the time being. I also changed the data being written o fields that are supposed to be boolean and write either "0" or "1" so maybe that has something to do with it as well. Will close the issue for the time being. Thanks again :)
